File: Tareas/Tarea03/TDAEjercicio03/listaligada.c
//
// Created by Jonthan Cordoba on 26/11/22.
//
#include <stdio.h>
#include <stdlib.h>
#include "listaligada.h"

Nodo * crearNodo(int x){
    Nodo * nuevo = NULL;
    nuevo = (Nodo *) malloc(sizeof(Nodo));
    if(nuevo == NULL) return NULL;
    nuevo->dato = x;
    nuevo->siguiente = NULL;

    return nuevo;
}

ListaLigada * listaLigada(){
    // Inicializamos la lista en NULL
    ListaLigada * lista = NULL;
    // Asignamos memoria para la lista
    lista = (ListaLigada *) malloc(sizeof(ListaLigada));
    if(lista == NULL) return NULL; // En caso de que no se le pueda asignar memoria
    lista->cabecera = NULL; // Asignamos valores

    return lista;
}

void insertar_inicio(ListaLigada * lista, int dato){
    Nodo * nuevo = crearNodo(dato);
    if(lista->cabecera == NULL){
        lista->cabecera = nuevo;
    } else{
        nuevo->siguiente = lista->cabecera;
        lista->cabecera = nuevo;
    }
}

void insertar_final(ListaLigada * lista, int dato){
    Nodo * nuevo = crearNodo(dato);
    if(lista->cabecera == NULL){
        lista->cabecera = nuevo;
    } else{
        Nodo * temp = lista->cabecera;
        while (temp->siguiente != NULL){
            temp = temp->siguiente;
        }
        temp->siguiente = nuevo;
    }
}

int no_nodos(ListaLigada * lista){
    if(lista->cabecera == 0){
        return 0;
    } else{
        Nodo * temp = lista->cabecera;
        int cont = 0;
        while(temp != NULL){
            temp = temp->siguiente;
            cont++;
        }
        return cont;
    }
}

void insertar_intermedio(ListaLigada * lista, int pos, int dato){
    Nodo * nuevo = crearNodo(dato);
    // Se toma el 0 como posición válida
    int noNodos = no_nodos(lista);
    if(pos<-1 || pos>noNodos+1){
        printf("Posición inválida\n");
        return;
    } else if(lista->cabecera == NULL) {
        lista->cabecera = nuevo;
    } else if(pos == 0){
        nuevo->siguiente = lista->cabecera;
        lista->cabecera = nuevo;
        return;
    } else{
        Nodo * temp = lista->cabecera;
        int i=0;
        while(i<pos-1){
            temp = temp->siguiente; // (pos-1)-th nodo
            i++;
        }
        nuevo->siguiente = temp->siguiente;
        temp->siguiente = nuevo;
    }
}

void borrar_inicio(ListaLigada * lista){
    if(lista->cabecera == NULL){
        printf("Lista Vacía!!\n");
    } else{
        Nodo * temp = lista->cabecera;
        lista->cabecera = lista->cabecera->siguiente;
        free(temp);
    }
}


void borrar_final(ListaLigada * lista) {
    Nodo *temp; // Nodo a borrar
    Nodo *prev; // Nodo previo al ultimo
    if (lista->cabecera == NULL) {
        printf("Lista Vacía!!\n");
    } else {
        temp = lista->cabecera;
        // Recorremos hasta el último nodo de la Lista
        while (temp->siguiente != NULL) {
            prev = temp;
            temp = temp->siguiente;
        }
        if (temp == lista->cabecera) {
            lista->cabecera = NULL;
        } else {
            /* Desconectar el enlace */
            prev->siguiente = NULL;
        }
        free(temp); // Borramos el ultimo
    }
}

void borrar_intermedio(ListaLigada * lista, int pos){
    if(lista->cabecera == NULL) {
        printf("Lista Vacía!!\n");
    } else {
        // Se toma el 0 como posición válida
        int noNodos = no_nodos(lista);
        if(pos<-1 || pos>noNodos){
            printf("Posición inválida\n");
            return;
        } else {
            Nodo * temp = NULL;
            if(pos == 0){
                temp = lista->cabecera;
                lista->cabecera = lista->cabecera->siguiente;
                free(temp);
                return;
            } else{
                int i=0;
                temp = lista->cabecera;
                while(i<pos-1){
                    temp = temp->siguiente; // (pos-1)-th nodo
                    i++;
                }
                Nodo * borrado = temp->siguiente; // (pos)-th nodo
                temp->siguiente = borrado->siguiente; // (pos+1)-th nodo
                free(borrado);
            }
        }
    }
}

void desplegar(ListaLigada * lista){
    Nodo * temp = lista->cabecera;
    if(temp == NULL){
        printf("Lista Vacía!!\n");
        return;
    }
    while (temp != NULL){
        printf("%d ", temp->dato);
        temp = temp->siguiente;
    }
    printf("\n");
}

void limpiarLista(ListaLigada * lista) {
    Nodo *actual = NULL;
    while (lista->cabecera != NULL) {
        actual = lista->cabecera;
        lista->cabecera = lista->cabecera->siguiente;
        free(actual);
    }
    free(lista);
}

File: Python/ListaLigada.py
#!/usr/bin/env python
# coding: utf-8

# In[9]:


class Nodo:
   def __init__(self, dato=None):
        self.dato = dato
        self.siguiente = None


# In[53]:


class ListaLigada:
    def __init__(self):
        self.cabecera = None
        
    def insertar_inicio(self, dato):
        nuevo = Nodo(dato)
        if self.cabecera is None:
            self.cabecera = nuevo
        else:
            nuevo.siguiente = self.cabecera
            self.cabecera = nuevo
    
    def insertar_final(self, dato):
        nuevo = Nodo(dato)
        if self.cabecera is None:
            self.cabecera = nuevo
        else:
            temp = self.cabecera
            while temp.siguiente is not None:
                temp = temp.siguiente
            temp.siguiente = nuevo
            
    def no_nodos(self):
        if self.cabecera is None:
            return 0;
        else:
            temp = self.cabecera
            cont = 0;
            while temp is not None:
                temp = temp.siguiente
                cont = cont+1
            return cont
    
    def insertar_intermedio(self, pos, dato):
        nuevo = Nodo(dato)
        noNodos = self.no_nodos()
        if (pos<-1) or (pos>noNodos+1):
            print("Posición Invalida")
            return
        elif self.cabecera is None:
            self.cabecera = nuevo
        elif (pos == 0):
            nuevo.siguiente = self.cabecera
            self.cabecera = nuevo
        else:
            temp = self.cabecera
            i = 0
            while (i<pos-1):
                temp = temp.siguiente
                i = i+1
            nuevo.siguiente = temp.siguiente
            temp.siguiente = nuevo
    
    def borrar_inicio(self):
        if self.cabecera is None:
            print("\nLista Vacia")
        else:
            temp = self.cabecera
            self.cabecera = self.cabecera.siguiente
            temp = None
    
    def borrar_final(self):
        if self.cabecera is None:
            print("\nLista Vacia")
        else:
            temp = self.cabecera
            while temp.siguiente is not None:
                prev = temp
                temp = temp.siguiente
            if temp is self.cabecera:
                self.cabecera = None
            else:
                prev.siguiente = None
            temp = None
    
    
    def borrar_intermedio(self, pos):
        if self.cabecera is None:
            print("\nLista Vacia")
        else:
            noNodos = self.no_nodos()
            if (pos<-1) or (pos>noNodos):
                print("Posición Invalida")
                return
            else:
                if (pos == 0):
                    temp = self.cabecera
                    self.cabecera = self.cabecera.siguiente
                    temp = None
                else:
                    i = 0
                    temp = self.cabecera
                    while (i<pos-1):
                        temp = temp.siguiente
                        i = i+1
                    borrado = temp.siguiente
                    temp.siguiente = borrado.siguiente
                    borrado = None
 
    def desplegar(self):
        temp = self.cabecera
        if temp is None:
            print("Lista Vacia")
        else:
            while temp is not None:
                print(str(temp.dato) + " ", end = " ")
                temp = temp.siguiente
            print("");


# In[54]:


lista = ListaLigada()
lista.insertar_inicio(3)
lista.insertar_inicio(5)
lista.insertar_inicio(4)
lista.insertar_inicio(23)
lista.insertar_inicio(18)
lista.desplegar()
lista.insertar_final(18)
lista.insertar_final(23)
lista.insertar_final(4)
lista.insertar_final(23)
lista.insertar_final(3)
lista.desplegar()
lista.insertar_intermedio(0, 29)
lista.insertar_intermedio(11, 29)
lista.insertar_intermedio(6, 29)
lista.desplegar()
n = lista.no_nodos()
print("Numero de nodos: " + str(n))
lista.borrar_inicio()
lista.borrar_inicio()
lista.borrar_inicio()
lista.desplegar()
lista.borrar_final()
lista.borrar_final()
lista.borrar_final()
lista.desplegar()
lista.borrar_intermedio(1)
lista.borrar_intermedio(5)
lista.borrar_intermedio(2)
lista.desplegar()
n = lista.no_nodos()
print("Numero de nodos: " + str(n))





